substrate/primitives/runtime/src/traits/transaction_extension/mod.rs

```diff
@@ -0,0 +1,603 @@
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError>;
+
+	/// Do any post-flight stuff for an extrinsic.
+	///
+	/// `_pre` contains the output of `prepare`.
+	///
+	/// This gets given the `DispatchResult` `_result` from the extrinsic and can, if desired,
+	/// introduce a `TransactionValidityError`, causing the block to become invalid for including
+	/// it.
+	///
+	/// On success, the caller must return the amount of unspent weight left over by this extension
+	/// after dispatch. By default, this function returns no unspent weight, which means the entire
+	/// weight computed for the worst case scenario is consumed.
+	///
+	/// WARNING: This function does not automatically keep track of accumulated "actual" weight.
+	/// Unless this weight is handled at the call site, use
+	/// [post_dispatch](TransactionExtension::post_dispatch)
+	/// instead.
+	///
+	/// Parameters:
+	/// - `pre`: `Self::Pre` returned by the result of the `prepare` call prior to dispatch.
+	/// - `info`: Information concerning, and inherent to, the transaction's call.
+	/// - `post_info`: Information concerning the dispatch of the transaction's call.
+	/// - `len`: The total length of the encoded transaction.
+	/// - `result`: The result of the dispatch.
+	///
+	/// WARNING: It is dangerous to return an error here. To do so will fundamentally invalidate the
+	/// transaction and any block that it is included in, causing the block author to not be
+	/// compensated for their work in validating the transaction or producing the block so far. It
+	/// can only be used safely when you *know* that the transaction is one that would only be
+	/// introduced by the current block author.
+	fn post_dispatch_details(
+		_pre: Self::Pre,
+		_info: &DispatchInfoOf<Call>,
+		_post_info: &PostDispatchInfoOf<Call>,
+		_len: usize,
+		_result: &DispatchResult,
+	) -> Result<Weight, TransactionValidityError> {
+		Ok(Weight::zero())
+	}
+
+	/// A wrapper for [`post_dispatch_details`](TransactionExtension::post_dispatch_details) that
+	/// refunds the unspent weight consumed by this extension into the post dispatch information.
+	///
+	/// If `post_dispatch_details` returns a non-zero unspent weight, which, by definition, must be
+	/// less than the worst case weight provided by [weight](TransactionExtension::weight), that
+	/// is the value refunded in `post_info`.
+	///
+	/// If no unspent weight is reported by `post_dispatch_details`, this function assumes the worst
+	/// case weight and does not refund anything.
+	///
+	/// For more information, look into
+	/// [post_dispatch_details](TransactionExtension::post_dispatch_details).
+	fn post_dispatch(
+		pre: Self::Pre,
+		info: &DispatchInfoOf<Call>,
+		post_info: &mut PostDispatchInfoOf<Call>,
+		len: usize,
+		result: &DispatchResult,
+	) -> Result<(), TransactionValidityError> {
+		let unspent_weight = Self::post_dispatch_details(pre, info, &post_info, len, result)?;
+		post_info.refund(unspent_weight);
+
+		Ok(())
+	}
```

Comment:
Generally not sure we need to have both of these functions. `post_dispatch` and `*_details`. This is not called that often that we can just do the `refund` call somewhere else.

Reply:
Only `post_dispatch` is supposed to be called by a user of the extension, as the docs state. This distinction is necessary if we want the weight refund to be visible from one extension's `post_dispatch` call to the next AND to have the refund done automatically by the extension trait logic and not explicitly by users.